$$\mathbf{Exercise-1}$$

Given function is:$$f(x)=400x_1^2+0.004x_2^4
$$
then the hassian matrix of the function will be:
$$\nabla^2f(x)=\begin{bmatrix}
800 &0\\
0& 0.048x_2^2
\end{bmatrix}$$

In [ ]:
import numpy as np

In [ ]:
def evalf(x):
  assert type(x) is np.ndarray 
  assert len(x)==2
  return 400*(x[0]**2)+0.004*(x[1]**4)

In [ ]:
def evalh(x):
  assert type(x) is np.ndarray
  assert len(x)==2
  return np.array([[800,0],[0,0.048*(x[1]**2)]])

In [ ]:
def evalg(x):
  assert type(x) is np.ndarray
  assert len(x)==2
  return np.array([800*x[0],0.016*(x[1]**3)])

In [ ]:
def d_k(x):
  assert type(x) is np.ndarray
  assert len(x) == 2
  if np.linalg.det(evalh(x)) == 0:
    raise ValueError('Inverse does not exists. Please check!')
  return np.linalg.inv(evalh(x))

In [ ]:
def compute_steplength_backtracking_scaled_direction(x, gradf, alpha_start, rho, gamma): #add appropriate arguments to the function 
  assert type(x) is np.ndarray and len(gradf) == 2 
  assert type(gradf) is np.ndarray and len(gradf) == 2 
  assert type(alpha_start) is float and alpha_start>=0. 
  assert type(rho) is float and rho>=0.
  assert type(gamma) is float and gamma>=0. 
  alpha = alpha_start
  p = - gradf
  D_k = d_k(x)
  r=rho
  y=gamma
  while evalf(x + alpha*np.matmul(D_k,p)) > evalf(x) + y*alpha* (np.matmul(np.matrix.transpose(gradf), np.matmul(D_k,p)) ):
    alpha=alpha*r
  return alpha

  
  
  #Complete the code 

In [ ]:

BACKTRACKING_LINE_SEARCH = 1
CONSTANT_STEP_LENGTH = 2

In [ ]:
#complete the code for gradient descent with scaling to find the minimizer

def find_minimizer_gdscaling(start_x, tol, line_search_type,*args):
  #Input: start_x is a numpy array of size 2, tol denotes the tolerance and is a positive float value
  assert type(start_x) is np.ndarray and len(start_x) == 2 #do not allow arbitrary arguments 
  assert type(tol) is float and tol>=0 
  x = start_x
  g_x = evalg(x)

  #initialization for backtracking line search
  if(line_search_type == BACKTRACKING_LINE_SEARCH):
    alpha_start = args[0]
    rho = args[1]
    gamma = args[2]
    #print('Params for Backtracking LS: alpha start:', alpha_start, 'rho:', rho,' gamma:', gamma)

  k = 0
  while (np.linalg.norm(g_x) > tol): #continue as long as the norm of gradient is not close to zero upto a tolerance tol
    D_k = d_k(x)
    if line_search_type == BACKTRACKING_LINE_SEARCH:
      
      step_length = compute_steplength_backtracking_scaled_direction(x, g_x, alpha_start, rho, gamma) #call the new function you wrote to compute the steplength
      #raise ValueError('BACKTRACKING LINE SEARCH NOT YET IMPLEMENTED')
    elif line_search_type == CONSTANT_STEP_LENGTH: #do a gradient descent with constant step length
      step_length = 1
    else:  
      raise ValueError('Line search type unknown. Please check!')
    
    #implement the gradient descent steps here  
    x = np.subtract(x, np.multiply(step_length,np.matmul(D_k, g_x))) #update x = x - step_length*g_x
    k += 1 #increment iteration
    g_x = evalg(x) #compute gradient at new point
  return x, k
  #Complete the code  

#Que.3


In [ ]:
my_start_x = np.array([2.0,2.0])
alpha_start = 1.0
rho = 0.5
gamma = 0.5
my_tol= 1e-9

x_opt_bls, k = find_minimizer_gdscaling(my_start_x, my_tol,CONSTANT_STEP_LENGTH)
#print(x_opt_bls)

print('\n\nminimizer of the function by using constant step length with scaling=',x_opt_bls)
print('the number of iterations using constant step length computation with scaling=',k)
print('minimum value of the function by using constant step length with scaling=',evalf(x_opt_bls))



minimizer of the function by using constant step length with scaling= [0.         0.00304488]
the number of iterations using constant step length computation with scaling= 16
minimum value of the function by using constant step length with scaling= 3.4382653805813626e-13


In [ ]:
my_start_x = np.array([2.0,2.0])
alpha_start = 1.0
rho = 0.5
gamma = 0.5
my_tol= 1e-9

x_opt_bls, k = find_minimizer_gdscaling(my_start_x, my_tol,BACKTRACKING_LINE_SEARCH,1.0,0.5,0.5)
#print(x_opt_bls)

print('\n\nminimizer of the function by using backtracking line search with scaling=',x_opt_bls)
print('the number of iterations using backtracking line search computation with scaling=',k)
print('minimum value of the function by using backtracking line search with scaling=',evalf(x_opt_bls))



minimizer of the function by using backtracking line search with scaling= [0.         0.00304488]
the number of iterations using backtracking line search computation with scaling= 16
minimum value of the function by using backtracking line search with scaling= 3.4382653805813626e-13


we can observe that the number of iterations(16 iterations) are same for both methods (newton's method(with backtracking line search) and newton's method(with constant step length)).


****
$\textbf{Constant step length line search in Newton's Method :}$

The value of Minimizer: $(0. ,        0.00304488)$

Minimum function value: $3.4382653805813626e-13 $

No. of Iterations: $16$
****
$\textbf{Backtracking line search with scaling in Newton's Method:}$

Minimizer: $(0. ,       0.00304488)$

Minimum function value: $3.4382653805813626e-13 $

No. of Iterations: $16$
****
The value of minimizer and the minimum value of the function is exactly same for both methods.

#Que.4

In [ ]:
def compute_steplength_backtracking(x, gradf, alpha_start, rho, gamma): #add appropriate arguments to the function 
  assert type(x) is np.ndarray and len(gradf) == 2 
  assert type(gradf) is np.ndarray and len(gradf) == 2 
  assert type(alpha_start) is float and alpha_start>=0. 
  assert type(rho) is float and rho>=0.
  assert type(gamma) is float and gamma>=0. 
  
  #Complete the code 
  alpha = alpha_start
  p=rho
  y=gamma
  #implement the backtracking line search
  while evalf(x+alpha*(-gradf)) > evalf(x)-y*alpha*np.dot((gradf.T),gradf):
    alpha=p*alpha


  #print('final step length:',alpha)
  return alpha
  

In [ ]:
def find_minimizer_gd(start_x, tol, line_search_type,*args):
  #Input: start_x is a numpy array of size 2, tol denotes the tolerance and is a positive float value
  assert type(start_x) is np.ndarray and len(start_x) == 2 #do not allow arbitrary arguments 
  assert type(tol) is float and tol>=0 
  x = start_x
  g_x = evalg(x)
  #initialization for backtracking line search
  if(line_search_type == BACKTRACKING_LINE_SEARCH):
    alpha_start = args[0]
    rho = args[1]
    gamma = args[2]
    print('Params for Backtracking LS: alpha start:', alpha_start, 'rho:', rho,' gamma:', gamma)

  k = 0
  #print('iter:',k, ' x:', x, ' f(x):', evalf(x), ' grad at x:', g_x, ' gradient norm:', np.linalg.norm(g_x))

  while (np.linalg.norm(g_x) > tol): #continue as long as the norm of gradient is not close to zero upto a tolerance tol
  

    if line_search_type == BACKTRACKING_LINE_SEARCH:
      step_length = compute_steplength_backtracking(x,g_x, alpha_start,rho, gamma) #call the new function you wrote to compute the steplength
      #raise ValueError('BACKTRACKING LINE SEARCH NOT YET IMPLEMENTED')
    elif line_search_type == CONSTANT_STEP_LENGTH: #do a gradient descent with constant step length
      step_length = 1
    else:  
      raise ValueError('Line search type unknown. Please check!')
    
    #implement the gradient descent steps here   
    x = np.subtract(x, np.multiply(step_length,g_x)) #update x = x - step_length*g_x
    k += 1 #increment iteration
    g_x = evalg(x) #compute gradient at new point

    #print('iter:',k, ' x:', x, ' f(x):', evalf(x), ' grad at x:', g_x, ' gradient norm:', np.linalg.norm(g_x))
  return x ,k  


In [ ]:
my_start_x = np.array([2.0,2.0])
alpha_start = 1.0
rho = 0.5
gamma = 0.5
my_tol= 1e-9
 
x_opt_bls, k= find_minimizer_gdscaling(my_start_x, my_tol, BACKTRACKING_LINE_SEARCH, 1.0, 0.5,0.5)
#print(x_opt_bls)
 
print('minimizer of the function by using backtracking line search with scaling=',x_opt_bls)
print('the number of iterations using backtracking line search computation with scaling=',k)
print('minimum value of the function by using backtracking line search with scaling=',evalf(x_opt_bls))

minimizer of the function by using backtracking line search with scaling= [0.         0.00304488]
the number of iterations using backtracking line search computation with scaling= 16
minimum value of the function by using backtracking line search with scaling= 3.4382653805813626e-13


In [ ]:
my_start_x = np.array([2.0,2.0])
alpha_start = 1.0
rho = 0.5
gamma = 0.5
my_tol= 1e-9

x_opt_bls, k = find_minimizer_gd(my_start_x, my_tol, BACKTRACKING_LINE_SEARCH, 1.0, 0.5,0.5)
#print(x_opt_bls)

print('minimizer of the function by using backtracking line search without scaling=',x_opt_bls)
print('the number of iterations using backtracking line search computation without scaling=',k)
print('minimum value of the function by using backtracking line search without scaling=',evalf(x_opt_bls))

Params for Backtracking LS: alpha start: 1.0 rho: 0.5  gamma: 0.5


KeyboardInterrupt: ignored

Above method is taking very long times means it takes billions of iterations to terminate.

$\textbf{Backtracking line search with scaling:}$

Minimizer: $[0.         ,0.00304488]$

Minimum function value: $ 3.4382653805813626e-13 $

No. of Iterations: $16$
****
gradient descent algorith with backtracking without scaling is taking very high number of iterations to terminate while backtraking with line search is taking only 16 iterations to terminate which is very low as compare to other.
****
$\textbf{Constant step length line search in Newton's Method :}$

The value of Minimizer: $(0. ,        0.00304488)$

Minimum function value: $3.4382653805813626e-13 $

No. of Iterations: $16$
***
here constant step length in newton's method is also taking 16 itertions to terminate which is very less as compared to backtracking without scaling
